xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/ClusterManager.java

```diff
@@ -0,0 +1,273 @@
+import com.linecorp.armeria.internal.common.util.ReentrantShortLock;
+import com.linecorp.armeria.xds.ClusterSnapshot;
+import com.linecorp.armeria.xds.ListenerRoot;
+import com.linecorp.armeria.xds.ListenerSnapshot;
+import com.linecorp.armeria.xds.RouteSnapshot;
+import com.linecorp.armeria.xds.SnapshotWatcher;
+import com.linecorp.armeria.xds.client.endpoint.ClusterManager.State;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class ClusterManager implements SnapshotWatcher<ListenerSnapshot>, AsyncCloseable,
+                                      Listenable<State>, Consumer<List<Endpoint>> {
+
+    private static final Logger logger = LoggerFactory.getLogger(ClusterManager.class);
+    private static final Map<ClusterSnapshot, ClusterEntry> INITIAL_CLUSTER_ENTRIES = new HashMap<>();
+
+    private final EventExecutor eventLoop;
+    private final SafeCloseable safeCloseable;
+    private final CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+
+    private volatile Map<ClusterSnapshot, ClusterEntry> clusterEntries = INITIAL_CLUSTER_ENTRIES;
+    @Nullable
+    private volatile ListenerSnapshot listenerSnapshot;
+    private final Set<CompletableFuture<?>> pendingRemovals = Sets.newConcurrentHashSet();
+    private boolean closed;
+
+    @GuardedBy("listenersLock")
+    private final List<Consumer<? super State>> listeners = new ArrayList<>();
+    private final ReentrantShortLock listenersLock = new ReentrantShortLock();
+
+    ClusterManager(ListenerRoot safeCloseable) {
+        eventLoop = safeCloseable.eventLoop();
+        this.safeCloseable = safeCloseable;
+        safeCloseable.addSnapshotWatcher(this);
+    }
+
+    ClusterManager(ClusterSnapshot clusterSnapshot) {
+        checkArgument(clusterSnapshot.endpointSnapshot() != null,
+                      "An endpoint snapshot must exist for the provided (%s)", clusterSnapshot);
+        eventLoop = CommonPools.workerGroup().next();
+        clusterEntries = ImmutableMap.of(clusterSnapshot, new ClusterEntry(clusterSnapshot, this));
+        safeCloseable = () -> {};
+    }
+
+    @Nullable
+    Endpoint selectNow(ClientRequestContext ctx) {
+        final Map<ClusterSnapshot, ClusterEntry> clusterEntries = this.clusterEntries;
+        for (Entry<ClusterSnapshot, ClusterEntry> entry: clusterEntries.entrySet()) {
+            // Just use the first snapshot for now
+            final ClusterEntry clusterEntry = entry.getValue();
+            return clusterEntry.selectNow(ctx);
+        }
+        return null;
+    }
+
+    @Override
+    public void snapshotUpdated(ListenerSnapshot listenerSnapshot) {
+        if (closed) {
+            return;
+        }
+        final RouteSnapshot routeSnapshot = listenerSnapshot.routeSnapshot();
+        if (routeSnapshot == null) {
+            return;
+        }
+
+        // it is important that the entries are added in order of ClusterSnapshot#index
+        // so that the first matching route is selected in #selectNow
```

Comment:
Q) Is the order preserved with `ImmutableMap.Builder`?

Reply:
That's the hope:

At least in the resource node, the list is indexed:
https://github.com/line/armeria/blob/b14f93461705093ef21ad4132e9ebb88bcb325f7/xds/src/main/java/com/linecorp/armeria/xds/RouteResourceNode.java#L107

And `ImmutableMap` supports insertion order:
https://github.com/google/guava/blob/aecd2f0ac00b50f7ff01964a41b20227775064f3/guava/src/com/google/common/collect/ImmutableMap.java#L402-L403

But let me know if I missed anything

Reply:
I didn't know that the fact. Thanks for the link.